File: apps/mobile/lib/features/memory/data/memory_service.dart
import 'dart:math' as math;

import 'package:trace_mobile/core/contracts/contracts.dart';
import 'package:uuid/uuid.dart';

final class MemoryService implements MemoryApi {
  MemoryService(this._objectStore, this._sightingStore, this._assetStore);

  final ObjectStore _objectStore;
  final SightingStore _sightingStore;
  final SecureAssetStore _assetStore;
  final _uuid = const Uuid();

  @override
  Future<TraceResult<FindLastSeenResponse>> findLastSeen(
    String objectId,
  ) async {
    final object = await _objectStore.get(objectId);
    if (object case TraceFailure<ObjectReference>(
      :final code,
      :final message,
    )) {
      return TraceFailure(code, message);
    }
    final latest = await _sightingStore.getLatest(objectId);
    if (latest case TraceFailure<Sighting?>(:final code, :final message)) {
      return TraceFailure(code, message);
    }
    return TraceSuccess(
      FindLastSeenResponse(
        objectId: objectId,
        tag: (object as TraceSuccess<ObjectReference>).value.tag,
        lastSeen: (latest as TraceSuccess<Sighting?>).value,
      ),
    );
  }

  @override
  Future<TraceResult<List<Sighting>>> getTimeline(
    String objectId, {
    int limit = 50,
  }) => _sightingStore.getTimeline(objectId, limit: limit);

  @override
  Future<TraceResult<RecordSightingResponse>> recordSighting(
    RecordSightingRequest request,
  ) async {
    final existing = await _sightingStore.getLatest(request.objectId);
    if (existing case TraceFailure<Sighting?>(:final code, :final message)) {
      return TraceFailure(code, message);
    }
    final latest = (existing as TraceSuccess<Sighting?>).value;
    final shouldDeduplicate =
        latest != null &&
        request.detectedAt.difference(latest.detectedAt).abs() <
            const Duration(minutes: 2) &&
        _nearby(latest.location, request.location);
    if (shouldDeduplicate) {
      final updated = Sighting(
        sightingId: latest.sightingId,
        objectId: latest.objectId,
        detectedAt: request.detectedAt,
        location: request.location ?? latest.location,
        confidence: math.max(latest.confidence, request.confidence),
        evidenceAssetId: latest.evidenceAssetId,
      );
      final result = await _sightingStore.update(updated);
      if (result case TraceFailure<Sighting>(:final code, :final message)) {
        return TraceFailure(code, message);
      }
      return TraceSuccess(
        RecordSightingResponse(
          sightingId: latest.sightingId,
          created: false,
          deduplicatedWith: latest.sightingId,
        ),
      );
    }

    String? evidenceAssetId;
    if (request.evidenceImage != null) {
      final asset = await _assetStore.write(
        type: SecureAssetType.sightingEvidence,
        plaintext: request.evidenceImage!.jpegBytes,
        mimeType: 'image/jpeg',
      );
      if (asset case TraceSuccess<SecureAsset>(:final value)) {
        evidenceAssetId = value.assetId;
      }
    }
    final sighting = Sighting(
      sightingId: _uuid.v4(),
      objectId: request.objectId,
      detectedAt: request.detectedAt,
      location: request.location,
      confidence: request.confidence,
      evidenceAssetId: evidenceAssetId,
    );
    final inserted = await _sightingStore.insert(sighting);
    if (inserted case TraceFailure<Sighting>(:final code, :final message)) {
      return TraceFailure(code, message);
    }
    return TraceSuccess(
      RecordSightingResponse(
        sightingId: sighting.sightingId,
        created: true,
        deduplicatedWith: null,
      ),
    );
  }

  bool _nearby(GeoFix? first, GeoFix? second) {
    if (first == null || second == null) return true;
    const earthRadiusMeters = 6371000.0;
    final lat1 = first.latitude * math.pi / 180;
    final lat2 = second.latitude * math.pi / 180;
    final deltaLat = (second.latitude - first.latitude) * math.pi / 180;
    final deltaLon = (second.longitude - first.longitude) * math.pi / 180;
    final a =
        math.sin(deltaLat / 2) * math.sin(deltaLat / 2) +
        math.cos(lat1) *
            math.cos(lat2) *
            math.sin(deltaLon / 2) *
            math.sin(deltaLon / 2);
    return earthRadiusMeters * 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a)) <=
        30;
  }
}

File: apps/android/feature/memory/src/main/kotlin/com/traceapp/feature/memory/MemoryService.kt
package com.traceapp.feature.memory

import com.traceapp.core.contracts.FindLastSeenResponse
import com.traceapp.core.contracts.GeoFix
import com.traceapp.core.contracts.MemoryApi
import com.traceapp.core.contracts.MostMatchedItem
import com.traceapp.core.contracts.ObjectStore
import com.traceapp.core.contracts.RecordSightingRequest
import com.traceapp.core.contracts.RecordSightingResponse
import com.traceapp.core.contracts.SecureAssetStore
import com.traceapp.core.contracts.SecureAssetType
import com.traceapp.core.contracts.Sighting
import com.traceapp.core.contracts.SightingStore
import com.traceapp.core.contracts.TraceError
import com.traceapp.core.contracts.TraceErrorCode
import com.traceapp.core.contracts.TraceResult
import java.util.UUID
import javax.inject.Inject
import javax.inject.Singleton
import kotlin.math.atan2
import kotlin.math.cos
import kotlin.math.sin
import kotlin.math.sqrt

@Singleton
class MemoryService @Inject constructor(
    private val objectStore: ObjectStore,
    private val sightingStore: SightingStore,
    private val assetStore: SecureAssetStore,
) : MemoryApi {
    override suspend fun recordSighting(
        request: RecordSightingRequest,
    ): TraceResult<RecordSightingResponse> {
        if (request.confidence !in 0f..1f) {
            return TraceResult.Failure(
                TraceError(TraceErrorCode.INVALID_INPUT, "Confidence must be between 0 and 1"),
            )
        }
        if (objectStore.get(request.objectId) is TraceResult.Failure) {
            return TraceResult.Failure(
                TraceError(TraceErrorCode.NOT_FOUND, "Object not found"),
            )
        }
        val latest = when (val result = sightingStore.getLatest(request.objectId)) {
            is TraceResult.Success -> result.value
            is TraceResult.Failure -> return result
        }
        if (latest != null && shouldDeduplicate(latest, request)) {
            val updated = latest.copy(
                detectedAtEpochMillis = request.detectedAtEpochMillis,
                location = request.location ?: latest.location,
                confidence = maxOf(latest.confidence, request.confidence),
            )
            return when (val result = sightingStore.update(updated)) {
                is TraceResult.Success -> TraceResult.Success(
                    RecordSightingResponse(latest.sightingId, false, latest.sightingId),
                )
                is TraceResult.Failure -> result
            }
        }

        val sightingId = UUID.randomUUID().toString()
        val evidenceAssetId = request.evidenceImage?.let { image ->
            when (
                val result = assetStore.write(
                    ownerRecordId = sightingId,
                    type = SecureAssetType.SIGHTING_EVIDENCE,
                    plaintext = image.jpegBytes,
                    mimeType = "image/jpeg",
                )
            ) {
                is TraceResult.Success -> result.value.assetId
                is TraceResult.Failure -> return result
            }
        }
        val sighting = Sighting(
            sightingId = sightingId,
            objectId = request.objectId,
            detectedAtEpochMillis = request.detectedAtEpochMillis,
            location = request.location,
            confidence = request.confidence,
            evidenceAssetId = evidenceAssetId,
        )
        return when (val result = sightingStore.insert(sighting)) {
            is TraceResult.Success -> TraceResult.Success(
                RecordSightingResponse(sighting.sightingId, true, null),
            )
            is TraceResult.Failure -> {
                evidenceAssetId?.let { assetStore.delete(it) }
                result
            }
        }
    }

    override suspend fun findLastSeen(objectId: String): TraceResult<FindLastSeenResponse> {
        val reference = when (val result = objectStore.get(objectId)) {
            is TraceResult.Success -> result.value
            is TraceResult.Failure -> return result
        }
        return when (val result = sightingStore.getLatest(objectId)) {
            is TraceResult.Success -> TraceResult.Success(
                FindLastSeenResponse(objectId, reference.tag, result.value),
            )
            is TraceResult.Failure -> result
        }
    }

    override suspend fun getTimeline(objectId: String, limit: Int): TraceResult<List<Sighting>> {
        if (limit !in 1..100) {
            return TraceResult.Failure(
                TraceError(TraceErrorCode.INVALID_INPUT, "Limit must be between 1 and 100"),
            )
        }
        return sightingStore.getTimeline(objectId, limit)
    }

    override suspend fun mostMatchedItem(atEpochMillis: Long): TraceResult<MostMatchedItem?> {
        val tagByObjectId = when (val result = objectStore.getAllReferences()) {
            is TraceResult.Success -> result.value.associate { it.objectId to it.tag }
            is TraceResult.Failure -> return result
        }
        val rows = when (val result = sightingStore.getAllSightingTimes()) {
            is TraceResult.Success -> result.value
            is TraceResult.Failure -> return result
        }
        return TraceResult.Success(
            UsageMatcher.topMatch(atEpochMillis, rows, tagByObjectId)?.let { match ->
                match.tag?.let { tag ->
                    MostMatchedItem(
                        objectId = match.objectId,
                        tag = tag,
                        similarity = match.similarity,
                        matchedEpochMillis = match.matchedEpochMillis,
                    )
                }
            },
        )
    }

    private fun shouldDeduplicate(latest: Sighting, request: RecordSightingRequest): Boolean {
        val timeDistance = request.detectedAtEpochMillis - latest.detectedAtEpochMillis
        if (timeDistance !in 0..DEDUPLICATION_WINDOW_MILLIS) return false
        val previousLocation = latest.location
        val currentLocation = request.location
        return if (previousLocation == null || currentLocation == null) {
            true
        } else {
            distanceMeters(previousLocation, currentLocation) < DEDUPLICATION_DISTANCE_METERS
        }
    }

    private fun distanceMeters(first: GeoFix, second: GeoFix): Double {
        val latitudeDelta = Math.toRadians(second.latitude - first.latitude)
        val longitudeDelta = Math.toRadians(second.longitude - first.longitude)
        val a = sin(latitudeDelta / 2) * sin(latitudeDelta / 2) +
            cos(Math.toRadians(first.latitude)) * cos(Math.toRadians(second.latitude)) *
            sin(longitudeDelta / 2) * sin(longitudeDelta / 2)
        return EARTH_RADIUS_METERS * 2 * atan2(sqrt(a), sqrt(1 - a))
    }

    private companion object {
        const val DEDUPLICATION_WINDOW_MILLIS = 120_000L
        const val DEDUPLICATION_DISTANCE_METERS = 30.0
        const val EARTH_RADIUS_METERS = 6_371_000.0
    }
}
